keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+tools in k/release must not depend on looking for a tag in the k/kubeadm
+repository when building kubeadm.
+
+The k/release tooling pulls out the k/kubeadm repo and checks out
+branch “release-x.yy” k/kubeadm build tooling must write a KUBE_VERSION
+file with static versions, but a dynamic build date/time.
+Kubeadm’s version information in the binary should be based on the
+KUBE_VERSION file.
+
+At this point the k/release tooling must tag the latest commit of this branch
+with a placeholder tag `pre-<version>`.
+
+###### k/k pushes a new tag
+
+The post-submit picks the change and based on the k/k tag version,
+checks out the according branch in k/kubeadm, finds the latest commit with
+`pre-<version>` tag and removes the tag. Then pushes the same `<version>` tag
+without the `pre-` prefix.
+
+This process needs to handle multiple tags present on the same commit.
+If there are multiple `pre-<version>` tag the post-submit job must only
+operate on the `<version>` that that matched the k/k tag.
+
+#### Dependency updates
+
+With the kubeadm moves out of k/k, the project will be decoupled
+from the /vendor and /staging dependencies that it currently has in k/k.
+This poses the question of how often the kubeadm project will have to
+update its k8s.io/ dependencies in the go.mod file.
+
+Example of a go.mod file for kubeadm:
+https://github.com/neolit123/kubeadm/blob/66b94f324b63005589bab7132193e8031001ba57/kubeadm/go.mod
+
+The current proposal is to manually update these dependencies right
+after each release or during a release cycle if something is critical.
+The k/k issue tracker and Slack will be the source of signal if something
+has to be updated.
+
+##### Scenarios:
+
+###### A bug in k8s.io/[some-library] is found and fixed in /staging or /vendor of k/k
+
+The kubeadm maintainers need to watch for such changes or bug reports
+and update the kubeadm go.mod file. Running the change in CI will confirm
+that it is not regressing.
+
+###### The libraries in k/k/staging have a lot of changes since the last k8s release
+
+The kubeadm maintainers need to pick commits to update the kubeadm go.mod
+file to. Running the changes in CI will confirm that they are not regressing.
+
+
+###### Kubernetes released with bug fixes in dependency X, that kubeadm did not include
+
+Kubeadm should consider backporting a dependency update depending on the
+severity of the issue.
+If the issue is low severity evaluate if the the backport can be skipped.
+If the issue is high servery consider a backport.
+
+### Risks and Mitigations
+
+#### A portion of the kubeadm backend changes and breaks users
+
+As mentioned in "User story 3", the kubeadm maintainers
+will give no guarantees about the kubeadm backend until it is
+announced as stable. README files in the kubeadm repository
+should denote that.
+
+It is out of scope for this proposal to try to solve the problem.
+
+#### The post-submit job for tagging/branching k/kubeadm fails
```

Comment:
While we need to get to an automated flow, I worry that we're jumping to a specific automation implementation without first thoroughly discussing and designing a process (even if initially operated manually) that covers the required functionality for versioning, build, and release.  It's especially problematic relative to the highly manual process today on k/k.  I'd like to see us talking more about requirements which would satisfy users needs first, and then process flow to implement that, and then automation of that.

Reply:
the post-submit job proposal came from test-infra / prow maintainers as the better recommended way to synchronize tags and branches between repositories.
the exact details of how it's going to work are not 100% clear.

i think, SIG Release should consider this as the experimental path for the first project breaking out of k/k without using k/k staging.

if the experiment is a failure, we can think of alternatives and fall back to manual tagging to suite release X. but there are currently no other good alternatives and i think we should proceed with the implementation.

if when kubectl moves out they decide to not use this mechanic and create something better kubeadm can start using theirs solution too.

>  I'd like to see us talking more about requirements which would satisfy users needs first, and then process flow to implement that, and then automation of that.

discussions have been ongoing for a while and i don't think an implementation detail like automatic-tagging and branching is something that needs more discussion.

Reply:
This experiment does emphasize the dire need for SIG Release to get k/release code into a state where it can be run to consume from arbitrary git remotes and output to arbitrary gs:// bucket.  Otherwise we'll be experimenting on production. :(*

Reply:
We can see how this works.  If we get a k/kubeadm repo ASAP, we can make make some tiny anago tweeks to try tag that repo, and/or and prow jobs to try to mirror tagging (skip branching for now for simpler experiment), and see how it goes.  Initially k/kubeadm might have some weird state this month or next, but we'll sort it out before declaring k/kubeadm is "production".